From 6056dfd56d66b387ba5c3271e56a35722b9c9d32 Mon Sep 17 00:00:00 2001
From: Wu Zhigang <zhigang.wu@linux.intel.com>
Date: Tue, 15 May 2018 17:14:53 +0800
Subject: [PATCH 1657/2302] ASoC: tdf8532: Fix the codec status error issue on
 APL-GPMRB

Based on the TDF8532 manual:
If the wait_state result is ok, we should send CLK_DISCONNECT
command to force codec from STANDBY(2) to IDLE(1).
If the wait_state result is timeout, the codec state should be
at Clockfail(7), we still should send CLK_DISCONNECT command
force the codec from Clockfail(7) to Idle(1).

Signed-off-by: Wu Zhigang <zhigang.wu@linux.intel.com>
Reviewed-by: Keyon Jie <yang.jie@linux.intel.com>
---
 sound/soc/codecs/tdf8532.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/sound/soc/codecs/tdf8532.c b/sound/soc/codecs/tdf8532.c
index 1860e82..c7bdb39 100644
--- a/sound/soc/codecs/tdf8532.c
+++ b/sound/soc/codecs/tdf8532.c
@@ -192,7 +192,7 @@ static int tdf8532_wait_state(struct tdf8532_priv *dev_data, u8 req_state,
 
 	ret = -ETIME;
 
-	dev_err(dev, "tdf8532-codec: state: %u, req_state: %u, ret: %d\n",
+	dev_warn(dev, "tdf8532-codec: state: %u, req_state: %u, ret: %d\n",
 			status_repl->state, req_state, ret);
 
 out:
@@ -228,8 +228,14 @@ static int tdf8532_stop_play(struct tdf8532_priv *tdf8532)
 		goto out;
 
 	ret = tdf8532_wait_state(tdf8532, STATE_STBY, ACK_TIMEOUT);
-	if (ret < 0)
-		goto out;
+
+	 /* Refer to TDF8532 manual:
+	  * If the wait_state result is ok, we should send CLK_DISCONNECT
+	  * command to force codec from STANDBY(2) to IDLE(1).
+	  * If the wait_state result is timeout, the codec state should be
+	  * at Clockfail(7), we still should send CLK_DISCONNECT command
+	  * force the codec from Clockfail(7) to Idle(1).
+	  */
 
 	ret = tdf8532_amp_write(tdf8532, SET_CLK_STATE, CLK_DISCONNECT);
 	if (ret < 0)
-- 
2.7.4

